src/detect-mqtt-reason-code.c

```diff
@@ -0,0 +1,329 @@
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].AppLayerTxMatch = DetectMQTTReasonCodeMatch;
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].Setup = DetectMQTTReasonCodeSetup;
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].Free  = DetectMQTTReasonCodeFree;
+    sigmatch_table[DETECT_AL_MQTT_REASON_CODE].RegisterTests = MQTTReasonCodeRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+
+    DetectAppLayerInspectEngineRegister("mqtt.reason_code",
+            ALPROTO_MQTT, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectMQTTReasonCodeGeneric);
+
+    mqtt_reason_code_id = DetectBufferTypeGetByName("mqtt.reason_code");
+}
+
+static int DetectEngineInspectMQTTReasonCodeGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
+                                          f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \internal
+ * \brief Function to match reason code of an MQTT 5.0 Tx
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into DetectMQTTReasonCodeData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectMQTTReasonCodeMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state,
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const uint8_t *de = (const uint8_t *)ctx;
+    uint8_t code;
+
+    if (!de)
+        return 0;
+
+    if (rs_mqtt_tx_get_reason_code(txv, &code) == 0) {
+        /* this function does not return a code that needs to be compared,
+           so we can just return the result of the check implemented in
+           Rust */
+        return rs_mqtt_tx_unsuback_has_reason_code(txv, *de);
+    } else {
+        if (code == *de)
+            return 1;
+    }
+    return 0;
+}
+
+/**
+ * \internal
+ * \brief This function is used to parse options passed via mqtt.reason_code: keyword
+ *
+ * \param rawstr Pointer to the user provided options
+ *
+ * \retval de pointer to DetectMQTTReasonCodeData on success
+ * \retval NULL on failure
+ */
+static uint8_t *DetectMQTTReasonCodeParse(const char *rawstr)
+{
+    uint8_t *de = NULL;
+#define MAX_SUBSTRINGS 30
+    int ret = 0;
+    uint8_t val;
+    int ov[MAX_SUBSTRINGS];
+
+    /* We need to check whether the input is numeric like this because
+       StringParseUint8() will silently parse numeric prefixes from
+       mixed alphanumeric input, which we don't want here */
```

Comment:
Hi, @satta !
Just an update: The fix for StringParse functions to work properly i.e. error out on trailing characters to numeric values was merged into master: https://github.com/OISF/suricata/commit/6f84515dd92502577f884dca3c4417c931ae4a16. You can rebase and remove this pcreexec call now.
Please let me know in case you have any issues. Thank you very much for your work! :)
L

Reply:
Great, thanks! :)